tests/integration_tests/functional/test_cpu_features.py

```diff
@@ -4,79 +4,226 @@
+    _tap, _, _ = test_microvm_with_ssh.ssh_network_config(network_config, '1')
+    test_microvm_with_ssh.start()
+
+    _check_cpu_topology(test_microvm_with_ssh, 1, 1, "0")
+    _check_cpu_features(test_microvm_with_ssh, 1, "false")
+    _check_cache_topology(test_microvm_with_ssh, 0, 0)
+
+
+def test_1vcpu_ht_enabled(test_microvm_with_ssh, network_config):
+    """Check the CPUID for a microvm with the specified config."""
+    test_microvm_with_ssh.spawn()
+    test_microvm_with_ssh.basic_config(vcpu_count=1, ht_enabled=True)
+    _tap, _, _ = test_microvm_with_ssh.ssh_network_config(network_config, '1')
+    test_microvm_with_ssh.start()
+
+    _check_cpu_topology(test_microvm_with_ssh, 1, 1, "0")
+    _check_cpu_features(test_microvm_with_ssh, 1, "false")
+    _check_cache_topology(test_microvm_with_ssh, 0, 0)
+
+
+def test_2vcpu_ht_disabled(test_microvm_with_ssh, network_config):
+    """Check the CPUID for a microvm with the specified config."""
+    test_microvm_with_ssh.spawn()
+    test_microvm_with_ssh.basic_config(vcpu_count=2, ht_enabled=False)
+    _tap, _, _ = test_microvm_with_ssh.ssh_network_config(network_config, '1')
+    test_microvm_with_ssh.start()
+
+    _check_cpu_topology(test_microvm_with_ssh, 2, 1, "0,1")
+    _check_cpu_features(test_microvm_with_ssh, 2, "true")
+    _check_cache_topology(test_microvm_with_ssh, 0, 1)
+
+
+def test_2vcpu_ht_enabled(test_microvm_with_ssh, network_config):
+    """Check the CPUID for a microvm with the specified config."""
+    test_microvm_with_ssh.spawn()
+    test_microvm_with_ssh.basic_config(vcpu_count=2, ht_enabled=True)
+    _tap, _, _ = test_microvm_with_ssh.ssh_network_config(network_config, '1')
+    test_microvm_with_ssh.start()
+
+    _check_cpu_topology(test_microvm_with_ssh, 2, 2, "0,1")
+    _check_cpu_features(test_microvm_with_ssh, 2, "true")
+    _check_cache_topology(test_microvm_with_ssh, 1, 1)
+
+
+def test_16vcpu_ht_disabled(test_microvm_with_ssh, network_config):
+    """Check the CPUID for a microvm with the specified config."""
+    test_microvm_with_ssh.spawn()
+    test_microvm_with_ssh.basic_config(vcpu_count=16, ht_enabled=False)
+    _tap, _, _ = test_microvm_with_ssh.ssh_network_config(network_config, '1')
+    test_microvm_with_ssh.start()
+
+    _check_cpu_topology(test_microvm_with_ssh, 16, 1, "0-15")
+    _check_cpu_features(test_microvm_with_ssh, 16, "true")
+    _check_cache_topology(test_microvm_with_ssh, 0, 15)
+
+
+def test_16vcpu_ht_enabled(test_microvm_with_ssh, network_config):
+    """Check the CPUID for a microvm with the specified config."""
+    test_microvm_with_ssh.spawn()
+    test_microvm_with_ssh.basic_config(vcpu_count=16, ht_enabled=True)
+    _tap, _, _ = test_microvm_with_ssh.ssh_network_config(network_config, '1')
+    test_microvm_with_ssh.start()
+
+    _check_cpu_topology(test_microvm_with_ssh, 16, 2, "0-15")
+    _check_cpu_features(test_microvm_with_ssh, 16, "true")
+    _check_cache_topology(test_microvm_with_ssh, 1, 15)
+
```

Comment:
All the tests above seem to start the microvm and then immediately attempt to run a command through SSH. Are they always succeeding? It looks like they would sometimes fail, if the guest network was up, but sshd hadn't started.

Reply:
Yes, they always work, but there was an interesting thing here. If instead of
```_tap, _, _ = test_microvm_with_ssh.ssh_network_config(network_config, '1')```
I would use
```test_microvm_with_ssh.ssh_network_config(network_config, '1')```

They would sometimes fail. Even though ```_tap``` isn't needed anywhere.

Maybe it's the problem that you mentioned. Is there any way to fix it ?

Reply:
What you are describing looks like a design issue we have within our `host_tools` code: `ssh_network_config()` returns a self-cleaning `Tap` object. I.e. when that object gets dropped, it will also delete its underlying host tap interface.

So, discarding the return value of `ssh_network_config()` means holding no references to the returned `Tap` object, causing it to get dropped and take the tap interface with it. If, on the other hand, you assign it to a dummy var `_tap`, that var will hold a reference to the `Tap` object, and keep it alive until it goes out of scope (in this case, until the end of the test function).